rosidl_generator_cpp/test/test_initialization.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include "rosidl_generator_cpp/msg/primitives_some_default.hpp"
+
+TEST(Test_initialization, no_arg_constructor) {
+  rosidl_generator_cpp::msg::PrimitivesSomeDefault def;
+
+  ASSERT_EQ(1.125f, def.float32_value);
+  ASSERT_EQ(2.4, def.float64_value);
+  ASSERT_EQ(0ULL, def.uint64_value);
+  ASSERT_EQ("bar", def.string_value);
+}
+
+TEST(Test_initialization, all_constructor) {
+  rosidl_generator_cpp::msg::PrimitivesSomeDefault def(
+    rosidl_generator_cpp::MessageInitialization::ALL);
+
+  ASSERT_EQ(1.125f, def.float32_value);
+  ASSERT_EQ(2.4, def.float64_value);
+  ASSERT_EQ(0ULL, def.uint64_value);
+  ASSERT_EQ("bar", def.string_value);
+}
+
+TEST(Test_initialization, zero_constructor) {
+  rosidl_generator_cpp::msg::PrimitivesSomeDefault def(
+    rosidl_generator_cpp::MessageInitialization::ZERO);
+
+  ASSERT_EQ(0.0f, def.float32_value);
+  ASSERT_EQ(0.0, def.float64_value);
+  ASSERT_EQ(0ULL, def.uint64_value);
+  ASSERT_EQ("", def.string_value);
+}
+
+TEST(Test_initialization, defaults_only_constructor) {
+  rosidl_generator_cpp::msg::PrimitivesSomeDefault def(
+    rosidl_generator_cpp::MessageInitialization::DEFAULTS_ONLY);
+
+  ASSERT_EQ(1.125f, def.float32_value);
+  ASSERT_EQ(2.4, def.float64_value);
+  // skip checking def.uint64_value because it doesn't have a default in this
+  // message and thus isn't initialized with DEFAULTS_ONLY
```

Comment:
It might be interesting to seed the memory with a known, non-zero, pattern and then do a placement new. Then you could check to make sure `uint64_value` is not `0`. Something like:

```c++
char * memory = new char[sizeof(rosidl_generator_cpp::msg::PrimitivesSomeDefault)];
ASSERT_NE(memory, nullptr);
std::memset(memory, 0xfe, sizeof(rosidl_generator_cpp::msg::PrimitivesSomeDefault));
rosidl_generator_cpp::msg::PrimitivesSomeDefault * def =
  new(memory) (rosidl_generator_cpp::MessageInitialization::DEFAULTS_ONLY);

ASSERT_NE(0ULL, def->uint64_value);

// This should be in a "scope exit" to prevent a memory leak during an assert raising...
def->~PrimitivesSomeDefault();
delete[] memory;
```

This might be fragile, because I don't know what the C++ standard says about this, it might let a compiler initialize it to zero if it wanted to do so during the placement new. But if you wanted to try and test this case, this would be how to do it I think.

Reply:
That's actually a really good idea, thanks @wjwwood I've gone ahead and implemented this as you pointed out, and it seems to work as advertised, at least on Linux.  I'll give it another CI job run on the other platforms to make sure that it works everywhere.

Reply:
If you don't run it now, I'd recommend running this test in a Windows debug CI job, because it's the only one that I know of that does something to the memory in order to check for uninitialized memory accesses. Though since you're doing memset before it _should_ be ok.

Reply:
Thanks for the heads-up.  I'll run a Windows Debug job to be sure.